source/extensions/filters/network/local_ratelimit/local_ratelimit.cc

```diff
@@ -10,26 +10,90 @@ namespace Extensions {
 namespace NetworkFilters {
 namespace LocalRateLimitFilter {
 
+SINGLETON_MANAGER_REGISTRATION(shared_local_ratelimit);
+
+SharedRateLimitSingleton::~SharedRateLimitSingleton() {
+  // Validate that all limiters were properly cleaned up when Configs are deleted.
+  ASSERT(limiters_.empty());
+}
+
+std::pair<LocalRateLimiterImplSharedPtr, const SharedRateLimitSingleton::Key*>
+SharedRateLimitSingleton::get(
+    const envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit& proto_config,
+    std::function<LocalRateLimiterImplSharedPtr()> create_fn) {
+  ASSERT_IS_MAIN_OR_TEST_THREAD();
+
+  if (proto_config.share_key().empty()) {
+    // If there is no key, never share.
+    return {create_fn(), nullptr};
+  }
+
+  LocalRateLimiterImplSharedPtr rate_limiter;
+  const Key key{proto_config.share_key(), proto_config.token_bucket()};
+  auto it = limiters_.find(key);
+  if (it == limiters_.end()) {
+    ENVOY_LOG(debug, fmt::format("LocalRateLimit for share_key '{}' is creating a new token bucket "
+                                 "due to no match found",
+                                 proto_config.share_key()));
+    rate_limiter = create_fn();
+    it = limiters_.emplace(key, rate_limiter).first;
+  } else {
+    ENVOY_LOG(
+        debug,
+        fmt::format("LocalRateLimit for share_key '{}' is using an existing matching token bucket",
+                    proto_config.share_key()));
+    rate_limiter = it->second.lock();
+    ASSERT(rate_limiter != nullptr);
+  }
+  return {rate_limiter, &it->first};
+}
+
+void SharedRateLimitSingleton::removeIfUnused(const Key* key) {
```

Comment:
Potentially add `ASSERT_IS_MAIN_OR_TEST_THREAD();` here also? I know we have had lots of bugs over the years with things accidentally getting run on other threads, and this would not be thread safe if so.